Segfault: kops replace -f spec.yaml --force if cluster doesn't exist yet

1. What `kops` version are you running? The command `kops version`, will display
 this information.
    1. kops-1.8.0-beta.1
1. What cloud provider are you using?
    1. AWS
1. What commands did you run?  What is the simplest way to reproduce this issue?

```
kops create $NAME
kops get $NAME -oyaml > spec.yaml
kops delete $NAME
kops replace -f spec.yaml --force
```

4. What happened after the commands executed?
    1. segfault
1. What did you expect to happen?
    1. cluster comes up
1. Please run the commands with most verbose logging by adding the `-v 10` flag.

```
kops replace -v10 -f ./manifests/kops/nameredacted --force
I1127 18:20:33.302057   11989 aws_utils.go:39] Querying EC2 for all valid regions
I1127 18:20:34.705499   11989 aws_cloud.go:948] Querying EC2 for all valid zones in region "us-east-1"
I1127 18:20:34.705665   11989 request_logger.go:45] AWS request: ec2/DescribeAvailabilityZones
I1127 18:20:34.983551   11989 status.go:57] Querying AWS for etcd volumes
I1127 18:20:34.983632   11989 status.go:68] Listing EC2 Volumes
I1127 18:20:34.983857   11989 request_logger.go:45] AWS request: ec2/DescribeVolumes
I1127 18:20:35.513892   11989 status.go:40] Cluster status (from cloud): {}
I1127 18:20:36.715943   11989 s3context.go:163] Found bucket "nameredacted-state" in region "us-east-1"
I1127 18:20:36.716009   11989 s3fs.go:176] Reading file "s3://nameredacted-state/nameredacted/config"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x2d8 pc=0x2149322]

goroutine 1 [running]:
k8s.io/kops/pkg/apis/kops/validation.ValidateClusterUpdate(0xc42012e400, 0xc420eee880, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/pkg/apis/kops/validation/cluster.go:39 +0x262
k8s.io/kops/pkg/client/simple/vfsclientset.(*ClusterVFS).Update(0xc420e320a0, 0xc42012e400, 0xc420eee880, 0xc420b5afa0, 0x3, 0x3)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/pkg/client/simple/vfsclientset/cluster.go:126 +0xd6
k8s.io/kops/pkg/client/simple/vfsclientset.(*VFSClientset).UpdateCluster(0xc420a96ec0, 0xc42012e400, 0xc420eee880, 0x0, 0x0, 0x0)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/pkg/client/simple/vfsclientset/clientset.go:50 +0x4d
main.RunReplace(0xc420a63a00, 0xc4204b38c0, 0x571b920, 0xc420092008, 0xc420aaad80, 0x1, 0xc42000e4f8)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/cmd/kops/replace.go:127 +0x5a1
main.NewCmdReplace.func1(0xc4204b38c0, 0xc420a7bb40, 0x0, 0x4)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/cmd/kops/replace.go:76 +0x74
k8s.io/kops/vendor/github.com/spf13/cobra.(*Command).execute(0xc4204b38c0, 0xc420a7bb00, 0x4, 0x4, 0xc4204b38c0, 0xc420a7bb00)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/vendor/github.com/spf13/cobra/command.go:603 +0x234
k8s.io/kops/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x5912860, 0x34f5600, 0x0, 0x0)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/vendor/github.com/spf13/cobra/command.go:689 +0x2fe
k8s.io/kops/vendor/github.com/spf13/cobra.(*Command).Execute(0x5912860, 0x5946670, 0x0)
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/vendor/github.com/spf13/cobra/command.go:648 +0x2b
main.Execute()
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/cmd/kops/root.go:95 +0x91
main.main()
	/private/tmp/kops-20171127-9784-1adfu3e/kops-1.8.0-beta.1/src/k8s.io/kops/cmd/kops/main.go:25 +0x20
```

7. Anything else do we need to know?
    1. I think this is the same bug as #3470, but manifesting in a workflow that is meant for primetime soon wrt #2404
    1. workaround is to `kops create $NAME` first

--- a/cmd/kops/replace.go
+++ b/cmd/kops/replace.go
@@ -23,12 +23,12 @@ import (
 
 	"github.com/golang/glog"
 	"github.com/spf13/cobra"
-	"k8s.io/kops/cmd/kops/util"
-	"k8s.io/kops/util/pkg/vfs"
-
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kops/cmd/kops/util"
 	kopsapi "k8s.io/kops/pkg/apis/kops"
 	"k8s.io/kops/pkg/kopscodecs"
+	"k8s.io/kops/util/pkg/vfs"
 	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
 	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
 	"k8s.io/kubernetes/pkg/kubectl/resource"
@@ -77,7 +77,7 @@ func NewCmdReplace(f *util.Factory, out io.Writer) *cobra.Command {
 		},
 	}
 	cmd.Flags().StringSliceVarP(&options.Filenames, "filename", "f", options.Filenames, "A list of one or more files separated by a comma.")
-	cmd.Flags().BoolVarP(&options.force, "force", "", false, "Force any changes, which will also create any non-existing respurce (defaults to instancegroups only)")
+	cmd.Flags().BoolVarP(&options.force, "force", "", false, "Force any changes, which will also create any non-existing resource")
 	cmd.MarkFlagRequired("filename")
 
 	return cmd
@@ -124,9 +124,29 @@ func RunReplace(f *util.Factory, cmd *cobra.Command, out io.Writer, c *replaceOp
 						return err
 					}
 
-					_, err = clientset.UpdateCluster(v, status)
+					// Check if the cluster exists already
+					clusterName := v.Name
+					cluster, err := clientset.GetCluster(clusterName)
 					if err != nil {
-						return fmt.Errorf("error replacing cluster: %v", err)
+						if errors.IsNotFound(err) {
+							cluster = nil
+						} else {
+							return fmt.Errorf("error fetching cluster %q: %v", clusterName, err)
+						}
+					}
+					if cluster == nil {
+						if !c.force {
+							return fmt.Errorf("cluster %v does not exist (try adding --force flag)", clusterName)
+						}
+						_, err = clientset.CreateCluster(v)
+						if err != nil {
+							return fmt.Errorf("error creating cluster: %v", err)
+						}
+					} else {
+						_, err = clientset.UpdateCluster(v, status)
+						if err != nil {
+							return fmt.Errorf("error replacing cluster: %v", err)
+						}
 					}
 				}
 
@@ -137,7 +157,11 @@ func RunReplace(f *util.Factory, cmd *cobra.Command, out io.Writer, c *replaceOp
 				}
 				cluster, err := clientset.GetCluster(clusterName)
 				if err != nil {
-					return fmt.Errorf("error fetching cluster %q: %v", clusterName, err)
+					if errors.IsNotFound(err) {
+						cluster = nil
+					} else {
+						return fmt.Errorf("error fetching cluster %q: %v", clusterName, err)
+					}
 				}
 				if cluster == nil {
 					return fmt.Errorf("cluster %q not found", clusterName)

--- a/docs/cli/kops_replace.md
+++ b/docs/cli/kops_replace.md
@@ -28,7 +28,7 @@ kops replace -f FILENAME
 
 ```
   -f, --filename stringSlice   A list of one or more files separated by a comma.
-      --force                  Force any changes, which will also create any non-existing respurce (defaults to instancegroups only)
+      --force                  Force any changes, which will also create any non-existing resource
 ```
 
 ### Options inherited from parent commands

--- a/pkg/client/simple/vfsclientset/cluster.go
+++ b/pkg/client/simple/vfsclientset/cluster.go
@@ -23,7 +23,9 @@ import (
 	"time"
 
 	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	"k8s.io/apimachinery/pkg/watch"
@@ -50,6 +52,7 @@ func (c *ClusterVFS) Get(name string, options metav1.GetOptions) (*api.Cluster, 
 	if options.ResourceVersion != "" {
 		return nil, fmt.Errorf("ResourceVersion not supported in ClusterVFS::Get")
 	}
+	// TODO: Return not found
 	return c.find(name)
 }
 
@@ -121,6 +124,10 @@ func (r *ClusterVFS) Update(c *api.Cluster, status *api.ClusterStatus) (*api.Clu
 	old, err := r.Get(clusterName, metav1.GetOptions{})
 	if err != nil {
 		return nil, err
+	}
+
+	if old == nil {
+		return nil, errors.NewNotFound(schema.GroupResource{Group: api.GroupName, Resource: "Cluster"}, clusterName)
 	}
 
 	if err := validation.ValidateClusterUpdate(c, status, old).ToAggregate(); err != nil {

